File: lib/utils/recommendation-migration.ts
/**
 * Recommendation Migration Utilities for Frontend
 * Handles migration from legacy recommendation formats to standardized FastAPI formats
 */

interface DiscountRange {
  min: number
  max: number
}

// Mapping from legacy recommendations to FastAPI standard
const LEGACY_TO_FASTAPI_MAP: Record<string, string> = {
  // Legacy format -> FastAPI standard
  immediate_action: 'discount_aggressive',
  high_priority: 'discount_aggressive',
  medium_priority: 'discount_moderate',
  discount_heavily: 'discount_aggressive',
  normal: 'maintain',

  // Keep existing FastAPI standards as-is
  dispose: 'dispose',
  donate: 'donate',  // NEW: AI scoring donation recommendation
  discount_aggressive: 'discount_aggressive',
  discount_moderate: 'discount_moderate',
  discount_light: 'discount_light',  // NEW: AI scoring light discount
  alert: 'alert',
  monitor: 'monitor',
  maintain: 'maintain',
}

// Display text mapping
const FASTAPI_TO_DISPLAY_MAP: Record<string, string> = {
  dispose: 'Dispose Immediately',
  donate: 'Donate to Charity',  // NEW: AI scoring donation recommendation
  discount_aggressive: 'Apply Heavy Discount',
  discount_moderate: 'Apply Moderate Discount',
  discount_light: 'Apply Light Discount',  // NEW: AI scoring light discount
  alert: 'Monitor Closely',
  monitor: 'Routine Monitoring',
  maintain: 'No Action Needed',
}

// Priority mapping for sorting (lower = higher priority)
const PRIORITY_ORDER: Record<string, number> = {
  dispose: 1,
  donate: 2,  // NEW: High priority for donation
  discount_aggressive: 3,
  discount_moderate: 4,
  discount_light: 5,  // NEW: Lower priority for light discounts
  alert: 6,
  monitor: 7,
  maintain: 8,
}

// Color mapping for UI
const COLOR_MAP: Record<string, string> = {
  dispose: 'destructive',
  donate: 'default',  // NEW: Neutral color for donation (positive action)
  discount_aggressive: 'destructive',
  discount_moderate: 'secondary',
  discount_light: 'secondary',  // NEW: Same as moderate discount
  alert: 'secondary',
  monitor: 'outline',
  maintain: 'outline',
}

/**
 * Migrate a legacy recommendation to FastAPI standard format
 */
export function migrateRecommendation(legacyRecommendation: string | null | undefined): string {
  if (!legacyRecommendation) {
    return 'maintain'
  }

  // Clean the input
  const cleaned = legacyRecommendation.trim().toLowerCase()

  // Check direct mapping
  if (cleaned in LEGACY_TO_FASTAPI_MAP) {
    return LEGACY_TO_FASTAPI_MAP[cleaned]
  }

  // Fuzzy matching for variations
  if (cleaned.includes('immediate') || cleaned.includes('urgent')) {
    return 'discount_aggressive'
  } else if (cleaned.includes('high') && cleaned.includes('priority')) {
    return 'discount_aggressive'
  } else if (cleaned.includes('medium') && cleaned.includes('priority')) {
    return 'discount_moderate'
  } else if (
    cleaned.includes('discount') &&
    (cleaned.includes('heavy') || cleaned.includes('heavily') || cleaned.includes('aggressive'))
  ) {
    return 'discount_aggressive'
  } else if (cleaned.includes('discount') && cleaned.includes('moderate')) {
    return 'discount_moderate'
  } else if (cleaned.includes('discount') && cleaned.includes('light')) {
    return 'discount_light'
  } else if (cleaned.includes('donate') || cleaned.includes('donation')) {
    return 'donate'
  } else if (cleaned.includes('dispose') || cleaned.includes('disposal')) {
    return 'dispose'
  } else if (cleaned.includes('normal') || cleaned.includes('maintain')) {
    return 'maintain'
  } else if (cleaned.includes('monitor')) {
    return 'monitor'
  } else if (cleaned.includes('alert')) {
    return 'alert'
  }

  // Default fallback
  return 'maintain'
}

/**
 * Get user-friendly display text for a recommendation
 */
export function getDisplayText(recommendation: string | null | undefined): string {
  // First migrate if it's legacy format
  const standardRecommendation = migrateRecommendation(recommendation)

  return (
    FASTAPI_TO_DISPLAY_MAP[standardRecommendation] ||
    standardRecommendation.replace(/_/g, ' ').replace(/\b\w/g, l => l.toUpperCase())
  )
}

/**
 * Get priority score for sorting (lower = higher priority)
 */
export function getPriorityScore(recommendation: string | null | undefined): number {
  const standardRecommendation = migrateRecommendation(recommendation)
  return PRIORITY_ORDER[standardRecommendation] || 99
}

/**
 * Get color variant for UI components
 */
export function getColorVariant(recommendation: string | null | undefined): string {
  const standardRecommendation = migrateRecommendation(recommendation)
  return COLOR_MAP[standardRecommendation] || 'outline'
}

/**
 * Categorize recommendation into action type
 */
export function getActionCategory(
  recommendation: string | null | undefined,
): 'critical' | 'action_needed' | 'monitor' | 'normal' {
  const standardRecommendation = migrateRecommendation(recommendation)

  if (standardRecommendation === 'dispose') {
    return 'critical'
  } else if (['donate', 'discount_aggressive'].includes(standardRecommendation)) {
    return 'action_needed'
  } else if (['discount_moderate', 'discount_light', 'alert'].includes(standardRecommendation)) {
    return 'monitor'
  } else {
    return 'normal'
  }
}

/**
 * Check if recommendation should show discount suggestions
 */
export function shouldShowDiscountSuggestion(recommendation: string | null | undefined): boolean {
  const standardRecommendation = migrateRecommendation(recommendation)
  return ['discount_aggressive', 'discount_moderate'].includes(standardRecommendation)
}

/**
 * Get suggested discount range based on recommendation and margin
 */
export function getSuggestedDiscountRange(
  recommendation: string | null | undefined,
  marginPercent: number,
): DiscountRange | null {
  const standardRecommendation = migrateRecommendation(recommendation)

  if (!shouldShowDiscountSuggestion(standardRecommendation)) {
    return null
  }

  // Calculate safe discount range based on margin
  const maxSafeDiscount = Math.min(Math.floor(marginPercent * 0.8), 60) // Don't go below 20% margin

  if (standardRecommendation === 'discount_aggressive') {
    return {
      min: Math.min(20, maxSafeDiscount),
      max: maxSafeDiscount,
    }
  } else if (standardRecommendation === 'discount_moderate') {
    return {
      min: 5,
      max: Math.min(25, maxSafeDiscount),
    }
  }

  return null
}

/**
 * Get action suggestions based on recommendation
 */
export function getActionSuggestions(recommendation: string | null | undefined): string[] {
  const standardRecommendation = migrateRecommendation(recommendation)

  const suggestions: Record<string, string[]> = {
    dispose: [
      'Remove from inventory immediately',
      'Check for similar expired batches',
      'Review disposal procedures',
    ],
    donate: [
      'Contact donation recipient',
      'Prepare for donation pickup',
      'Ensure food safety compliance',
      'Document donation for tax benefits',
    ],
    discount_aggressive: [
      'Apply 25-50% discount',
      'Move to prominent display area',
      'Create bundle offers',
      'Notify customers via app/email',
    ],
    discount_moderate: [
      'Apply 10-25% discount',
      'Feature in promotional materials',
      'Monitor daily for changes',
    ],
    discount_light: [
      'Apply 5-15% discount',
      'Feature in regular promotions',
      'Monitor for changes',
    ],
    alert: [
      'Monitor closely for next 24-48 hours',
      'Prepare discount strategy',
      'Check similar products',
    ],
    monitor: ['Continue routine monitoring', 'Review weekly for changes'],
    maintain: ['No immediate action required', 'Continue normal operations'],
  }

  return suggestions[standardRecommendation] || ['Continue monitoring']
}

File: lifo_api/app/utils/recommendation_migration.py
"""
Recommendation Migration Utilities
Handles migration from legacy recommendation formats to standardized FastAPI formats
"""



class RecommendationMigrator:
    """Handles mapping between legacy and FastAPI recommendation formats"""

    # Mapping from legacy recommendations to FastAPI standard
    LEGACY_TO_FASTAPI_MAP: dict[str, str] = {
        # Legacy format -> FastAPI standard
        "immediate_action": "discount_aggressive",
        "high_priority": "discount_aggressive",
        "medium_priority": "discount_moderate",
        "discount_heavily": "discount_aggressive",
        "normal": "maintain",

        # Keep existing FastAPI standards as-is
        "dispose": "dispose",
        "discount_aggressive": "discount_aggressive",
        "discount_moderate": "discount_moderate",
        "alert": "alert",
        "monitor": "monitor",
        "maintain": "maintain",
    }

    # Reverse mapping for compatibility
    FASTAPI_TO_DISPLAY_MAP: dict[str, str] = {
        "dispose": "Dispose Immediately",
        "discount_aggressive": "Apply Heavy Discount",
        "discount_moderate": "Apply Moderate Discount",
        "alert": "Monitor Closely",
        "monitor": "Routine Monitoring",
        "maintain": "No Action Needed",
    }

    # Priority mapping for sorting
    PRIORITY_ORDER: dict[str, int] = {
        "dispose": 1,
        "discount_aggressive": 2,
        "discount_moderate": 3,
        "alert": 4,
        "monitor": 5,
        "maintain": 6,
    }

    @classmethod
    def migrate_recommendation(cls, legacy_recommendation: str) -> str:
        """
        Migrate a legacy recommendation to FastAPI standard format
        
        Args:
            legacy_recommendation: The legacy recommendation value
            
        Returns:
            str: Standardized FastAPI recommendation
        """
        if not legacy_recommendation:
            return "maintain"

        # Clean the input
        cleaned = legacy_recommendation.strip().lower()

        # Check direct mapping
        if cleaned in cls.LEGACY_TO_FASTAPI_MAP:
            return cls.LEGACY_TO_FASTAPI_MAP[cleaned]

        # Fuzzy matching for variations
        if "immediate" in cleaned or "urgent" in cleaned:
            return "discount_aggressive"
        elif "high" in cleaned and "priority" in cleaned:
            return "discount_aggressive"
        elif "medium" in cleaned and "priority" in cleaned:
            return "discount_moderate"
        elif "discount" in cleaned and ("heavy" in cleaned or "heavily" in cleaned):
            return "discount_aggressive"
        elif "discount" in cleaned and "moderate" in cleaned:
            return "discount_moderate"
        elif "normal" in cleaned or "maintain" in cleaned:
            return "maintain"
        elif "monitor" in cleaned:
            return "monitor"
        elif "alert" in cleaned:
            return "alert"
        elif "dispose" in cleaned:
            return "dispose"

        # Default fallback
        return "maintain"

    @classmethod
    def get_display_text(cls, recommendation: str) -> str:
        """
        Get user-friendly display text for a recommendation
        
        Args:
            recommendation: FastAPI standard recommendation
            
        Returns:
            str: Human-readable display text
        """
        # First migrate if it's legacy format
        standard_recommendation = cls.migrate_recommendation(recommendation)

        return cls.FASTAPI_TO_DISPLAY_MAP.get(
            standard_recommendation,
            recommendation.replace("_", " ").title()
        )

    @classmethod
    def get_priority_score(cls, recommendation: str) -> int:
        """
        Get priority score for sorting (lower = higher priority)
        
        Args:
            recommendation: Recommendation to score
            
        Returns:
            int: Priority score (1 = highest priority)
        """
        # First migrate if it's legacy format
        standard_recommendation = cls.migrate_recommendation(recommendation)

        return cls.PRIORITY_ORDER.get(standard_recommendation, 99)

    @classmethod
    def get_action_category(cls, recommendation: str) -> str:
        """
        Categorize recommendation into action type
        
        Args:
            recommendation: Recommendation to categorize
            
        Returns:
            str: Action category (critical, action_needed, monitor, normal)
        """
        standard_recommendation = cls.migrate_recommendation(recommendation)

        if standard_recommendation in ["dispose"]:
            return "critical"
        elif standard_recommendation in ["discount_aggressive"]:
            return "action_needed"
        elif standard_recommendation in ["discount_moderate", "alert"]:
            return "monitor"
        else:
            return "normal"

    @classmethod
    def should_show_discount_suggestion(cls, recommendation: str) -> bool:
        """
        Check if recommendation should show discount suggestions
        
        Args:
            recommendation: Recommendation to check
            
        Returns:
            bool: True if discount suggestions should be shown
        """
        standard_recommendation = cls.migrate_recommendation(recommendation)
        return standard_recommendation in ["discount_aggressive", "discount_moderate"]

    @classmethod
    def get_suggested_discount_range(cls, recommendation: str, margin_percent: float) -> dict[str, int] | None:
        """
        Get suggested discount range based on recommendation and margin
        
        Args:
            recommendation: FastAPI recommendation
            margin_percent: Product margin percentage
            
        Returns:
            Optional[Dict[str, int]]: Dict with 'min' and 'max' discount percentages
        """
        standard_recommendation = cls.migrate_recommendation(recommendation)

        if not cls.should_show_discount_suggestion(standard_recommendation):
            return None

        # Calculate safe discount range based on margin
        max_safe_discount = min(int(margin_percent * 0.8), 60)  # Don't go below 20% margin

        if standard_recommendation == "discount_aggressive":
            return {
                "min": min(20, max_safe_discount),
                "max": max_safe_discount
            }
        elif standard_recommendation == "discount_moderate":
            return {
                "min": 5,
                "max": min(25, max_safe_discount)
            }

        return None


# Utility functions for easy importing
def migrate_recommendation(recommendation: str) -> str:
    """Convenience function for migrating recommendations"""
    return RecommendationMigrator.migrate_recommendation(recommendation)


def get_display_text(recommendation: str) -> str:
    """Convenience function for getting display text"""
    return RecommendationMigrator.get_display_text(recommendation)


def get_priority_score(recommendation: str) -> int:
    """Convenience function for getting priority score"""
    return RecommendationMigrator.get_priority_score(recommendation)
